inception-kb/src/main/java/de/tudarmstadt/ukp/inception/kb/reification/WikiDataReification.java

```diff
@@ -388,6 +413,17 @@ public void upsertQualifier(KnowledgeBase kb, KBQualifier aQualifier)
         });
     }
 
+    private int qualifierIndex(KBQualifier aQualifier)
+    {
+        List<KBQualifier> qualifierList = aQualifier.getKbStatement().getQualifiers();
+        for (KBQualifier qualifier : qualifierList) {
+            if (qualifier.getOriginalStatements().equals(aQualifier.getOriginalStatements())) {
```

Comment:
KBQualifier.equals() method is a good idea, index as a return value?

Reply:
If a equals method is added, mind also a hashCode method must be added. The best idea is usually to auto-generate these methods via the IDE because they must be in sync (i.e. based on the same fields, handle inheritance in the same way, etc.)

The qualifier would not know its index, so: no. Equals must compare two objects based on properties which uniquely identify them. Here e.g. by comparing their original statements - but comparing them in such a way that the comparison does *not* depend on the order of the statements in the list. This could be done by sorting the list prior to the comparison (if that makes sense) or by using a order-independent comparison algorithm.

Reply:
You also could just use a Set instead of a list, because a list implies order and possible duplicates.